Mathlib/AlgebraicGeometry/PrimeSpectrum/ConstructibleSet.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2024 Yaël Dillies, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.PrimeSpectrum.Basic
+import Mathlib.Order.SuccPred.WithBot
+
+/-!
+# Constructible sets in the prime spectrum
+
+This file provides tooling for manipulating constructible sets in the prime spectrum of a ring.
+
+## TODO
+
+Link to constructible sets in a topological space.
+-/
+
+open Finset
+open scoped Polynomial
+
+namespace PrimeSpectrum
+variable {R S T : Type*} [CommRing R] [CommRing S] [CommRing T]
+
+variable (R) in
+/-- The data of a constructible set `s` is finitely many tuples `(f, g₁, ..., gₙ)` such that
+`s = ⋃ (f, g₁, ..., gₙ), V(g₁, ..., gₙ) \ V(f)`.
+
+To obtain `s` from its data, use `PrimeSpectrum.ConstructibleSetData.toSet`. -/
+abbrev ConstructibleSetData := Finset (Σ n, R × (Fin n → R))
+
+namespace ConstructibleSetData
+
+/-- Given the data of the constructible set `s`, build the data of the constructible set
+`{I | {x | f x ∈ I} ∈ s}`. -/
+def map [DecidableEq S] (f : R →+* S) (s : ConstructibleSetData R) : ConstructibleSetData S :=
+  s.image <| Sigma.map id fun _n (r, g) ↦ (f r, f ∘ g)
+
+@[simp] lemma map_id [DecidableEq R] (s : ConstructibleSetData R) : s.map (.id _) = s := by
+  unfold map Sigma.map; simp [RingHom.coe_id]
+
+lemma map_comp [DecidableEq S] [DecidableEq T] (f : S →+* T) (g : R →+* S)
+    (s : ConstructibleSetData R) : s.map (f.comp g) = (s.map g).map f := by
+  unfold map Sigma.map; simp [image_image, Function.comp_def]
+
+/-- Given the data of a constructible set `s`, namely finitely many tuples `(f, g₁, ..., gₙ)` such
+that `s = ⋃ (f, g₁, ..., gₙ), V(g₁, ..., gₙ) \ V(f)`, return `s`. -/
+def toSet (S : ConstructibleSetData R) : Set (PrimeSpectrum R) :=
+  ⋃ x ∈ S, zeroLocus (Set.range x.2.2) \ zeroLocus {x.2.1}
```

Comment:
I wonder why you can restrict to finite many `g`s? Are you only dealing with Noetherian rings or ...?

Reply:
Oh I see, if the whole space is compact then any retrocompact set is compact.